resolwe_bio/processes/clustering/hierarchical_clustering.py

```diff
@@ -0,0 +1,899 @@
+    df = pd.read_csv(
+        filepath_or_buffer=fname,
+        sep=sep,
+        header=0,
+        index_col=0,
+        compression="gzip",
+        dtype={
+            0: str,
+            1: float,
+        },
+        keep_default_na=False,
+    )
+    df.index = df.index.map(str)
+    if not gene_set:
+        return df
+    intersection = [gene for gene in gene_set if gene in df.index]
+    return df.loc[intersection]
+
+
+def get_expressions(fnames, sep="\t", gene_set=[]):
+    """Read expressions from files.
+
+    Return only expressions of genes that are listed in all samples and in gene_set.
+
+    """
+    dfs = [get_expression(fname, sep=sep, gene_set=gene_set) for fname in fnames]
+    inner = pd.concat(dfs, axis=1, join="inner")
+    outer = pd.concat(dfs, axis=1, join="outer", sort=True)
+    if gene_set:
+        excluded = sorted(set(gene_set).difference(set(inner.index)))
+    else:
+        excluded = sorted(outer.index.difference(inner.index))
+    return inner, excluded
+
+
+def transform(expressions, log2=False, const=1.0, z_score=False, ddof=1):
+    """Compute log2 and normalize expression values.
+
+    Parameters:
+    - log2: use log2(x+const) transformation
+    - const: an additive constant used in computation of log2
+    - z_score: use Z-score normalization
+    - ddof: degrees of freedom used in computation of Z-score
+
+    """
+    if log2:
+        expressions = expressions.applymap(lambda x: np.log2(x + const))
+        if expressions.isnull().values.any():
+            msg = "Cannot apply log2 to expression values."
+            return msg
+    if z_score:
+        expressions = expressions.apply(
+            lambda x: zscore(x, ddof=ddof), axis=1, result_type="broadcast"
+        )
+        expressions.fillna(value=0.0, inplace=True)
+    return expressions
+
+
+def get_distance_metric(distance_metric):
+    """Get distance metric."""
+    if distance_metric == "spearman":
+        return lambda x, y: 1.0 - spearmanr(x, y).correlation
+    elif distance_metric == "pearson":
+        return "correlation"
+    return distance_metric
+
+
+def is_const(values):
+    """Return True, if all values are approximately equal, otherwise return False."""
+    mn = np.min(values)
+    mx = np.max(values)
+    if mn + mx == 0.0:
+        return mn == mx
+    else:
+        return (mx - mn) / abs(mx + mn) < 1.0e-6
+
+
+def remove_const_samples(expressions):
+    """Remove samples with constant expression profile across genes."""
+    matches = expressions.apply(lambda x: not is_const(x), axis=0)
+    return expressions.loc[:, matches], matches.values.tolist()
+
+
+def remove_const_genes(expressions):
+    """Remove genes with constant expression profile across samples."""
+    matches = expressions.apply(lambda x: not is_const(x), axis=1)
+    return expressions.loc[matches], matches[~matches].index.tolist()
+
+
+def output_json(result=dict(), fname=None):
```

Comment:
It is best to retain only the `fname` interface here as the `print()` option is I think not even supported anymore in Python-based processes